### DISCUSSION TITLE
Best practices for organizing a big hairy state machine??

### DISCUSSION THREAD

**Discussion by HerbCaudill** [statelyai/xstate#4783] [Q&A] [answered]
I just migrated a big state machine from v4 to v5. It's part of the [@localfirst/auth](http://github.com/local-first-web/auth) library and implements a connection protocol for authenticating communication between peers in a local-first setting.   

There's a class wrapping the state machine that gives it network access and a public API. 

This class was already an ugly beast before the migration, but at least before I had the state machine config in one file and the implementation in another. 

Now I've inlined the config in order to get the strong-typing benefits of the new `setup(...).createMachine(...)` API. 

I've spent a few hours now cleaning it up to try to turn it into something that won't cause my fellow developers to reconsider their career choices, but there doesn't seem to be any way around the fact that I now have 649 lines of code in my constructor alone. 🤡

You can [take a look here](https://github.com/local-first-web/auth/blob/d2d9b3c9209f869f8c009b528c92d3fc88a38663/packages/auth/src/connection/Connection.ts#L105-L703), but you might need a stiff drink first. Here's a redacted summary of the ugly facts:

```ts
const machine = setup({
  types: {
    context: {} as ConnectionContext,
    events: {} as ConnectionMessage,
  },
  actions: {
    //... 307 lines of code
  },
  guards: {
    // ... 45 lines of code
}).createMachine({
  //... 239 lines of code
})
```

So ... what's my question. I guess I'm wondering 

- is this ✌️normal✌️?
- is there a way to carve this up into smaller pieces and still get those TypeScript jaggies when I err, which I do sincerely love?
- or should I be, I don't know, breaking this down into multiple smaller state machines? It seems to me like there's just one actor here, so I'm not sure how that would work, but maybe I'm not looking at in the right way. 
- or ... what?

Thanks!!

**Comment by davidkpiano** [comment, marked as answer]
This is normal! Sure, the code seems like a lot, but when you visualize it (currently requires a pro subscription): https://github.stately.ai/local-first-web/auth/blob/d2d9b3c9209f869f8c009b528c92d3fc88a38663/packages/auth/src/connection/Connection.ts#L105-L703 - it actually becomes easily understandable. I got the gist of the state machine in under 30 seconds, just looking at the diagram:

<img width="668" alt="CleanShot 2024-03-04 at 17 37 56@2x" src="https://github.com/statelyai/xstate/assets/1093738/0575f8cb-11c9-44c6-9643-7bdc0b07e61a">

**Reply by HerbCaudill** [reply]
![image](https://github.com/statelyai/xstate/assets/2136620/53736bed-1d95-45b2-af14-a46ad545770b)

**Reply by with-heart** [reply]
@HerbCaudill I like you

**Comment by HerbCaudill** [comment]
see also: follow-up/rephrasing in #4812

**Comment by Bonobomagno** [comment]
Is this still a thing? Besides the "pro subscription", the old inspector is not usable and the code of big machines is really really too big. Splitting works, but the type system broke, and you need to spam a bunch of any or expect-error.

**Reply by Bonobomagno** [reply]
Nevermind, createStateConfig is there, I will give a try
